components/common/src/lib.rs

```diff
@@ -202,15 +203,46 @@ pub mod sync {
         pub fn read(&self) -> ReadGuard<T> { self.inner.read_recursive() }
 
         #[cfg(feature = "lock_as_mutex")]
-        pub fn read(&self) -> ReadGuard<T> { self.inner.lock() }
+        pub fn read(&self) -> ReadGuard<T> {
+            use parking_lot::deadlock;
+            use std::thread;
+
+            if let Some(guard) = self.inner.try_lock_for(Duration::from_secs(5)) {
+                guard
+            } else {
+                let detector = thread::spawn(move || {
+                    loop {
+                        thread::sleep(Duration::from_secs(1));
+                        let deadlocks = deadlock::check_deadlock();
+                        if deadlocks.is_empty() {
+                            continue;
+                        }
+
```

Comment:
Is it possible for `deadlock::check_deadlock()` to somehow return empty iterators for a long time?

It seems like if it could, then we'd wait a long time to join the detector thread. 

Since we're dealing with a 5 second window, and checking once a second, could we ensure that the thread will always end by exiting once we've done 5 deadlock checks? I imagine it'd be highly unlikely we'd ever get there, but if we actually did exit this code by that path, that would be interesting by itself.

Reply:
Excellent catch! This code is busted. Will fix it up.